### PR TITLE
Bignum: Implement fixed width modular negation

### PR DESCRIPTION
**Pull request by minosgalanakis**
## Description

This pr is importing the modular negation logic from Hanno's prototype and is adding standard tests for it.

# Gatekeeper checklist

- [x] **changelog**  not required
- [x] **backport** not required
- [x] **tests** provided

### CODE REVIEW

File: library/bignum_mod_raw.h
Comment (tom-cosgrove-arm) [inline library/bignum_mod_raw.h:245]: I don't like the phrasing "must have enough limbs". mod_raw should be working with fixed-size MPIs.

For example: if X has more limbs than A, only the number of limbs of A are set, so if X is then treated as a number with more limbs than A the result will be wrong.

Therefore I think we should be saying "A and X must have the same number of limbs as the modulus".

Hopefully the work that @wernerlewis is doing on the mod_raw conventions will help here.

File: library/bignum_mod_raw.h
Comment (tom-cosgrove-arm) [inline library/bignum_mod_raw.h:248]: I think this can just be "The modulus to use."

File: scripts/mbedtls_dev/bignum_mod_raw.py
Comment (tom-cosgrove-arm) [inline scripts/mbedtls_dev/bignum_mod_raw.py:117]: Shouldn't the symbol just be `-`?

Comment (minosgalanakis) [reply scripts/mbedtls_dev/bignum_mod_raw.py:117]: I was trying to not replicate the symbol of subtraction. This is for debugging purposes only anyhow.

Comment (tom-cosgrove-arm) [reply scripts/mbedtls_dev/bignum_mod_raw.py:117]: Well, it's what goes into the test case. Since `-` is the symbol for negation, and the test name includes "modular", I don't think there's any risk of confusion

File: tests/suites/test_suite_bignum_mod_raw.function
Comment (tom-cosgrove-arm) [inline tests/suites/test_suite_bignum_mod_raw.function:573]: Could also do this the other way round

```suggestion
    mbedtls_mpi_mod_raw_neg( A, A, &m );
    ASSERT_COMPARE( A, bytes, X, bytes );
```

Comment (minosgalanakis) [reply tests/suites/test_suite_bignum_mod_raw.function:573]: I considered that, was not sure if I should attemp to modify A, should other further testing it needed. But as long as this is the last test this will work. 

Will change it on next push.

File: tests/suites/test_suite_bignum_mod_raw.function
Comment (tom-cosgrove-arm) [inline tests/suites/test_suite_bignum_mod_raw.function:549]: Err, these were the places that the extra spaces stood out, not the ones above - sorry

File: tests/suites/test_suite_bignum_mod_raw.function
Comment (tom-cosgrove-arm) [inline tests/suites/test_suite_bignum_mod_raw.function:298]: these extra spaces could probably have been left, sorry

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM

File: library/bignum_mod_raw.h
Comment (tom-cosgrove-arm) [inline library/bignum_mod_raw.h:242]:
```suggestion
 * The size of the operation is determined by \p N: both \X and \A
 * must have the same number of limbs as \p N.
 *
 * \p X may be aliased to \p A.
```

**Comment by tom-cosgrove-arm** [comment]
So there's a small documentation issue that I'd rather we fixed before merging; everything else looks good now

File: library/bignum_mod_raw.h
Comment (minosgalanakis) [reply library/bignum_mod_raw.h]: Ammended the documentation commit to include that suggestion

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

File: library/bignum_mod_raw.c
Comment (gabor-mezei-arm) [inline library/bignum_mod_raw.c:188]: Should be called `N` to follow the naming convention.

**Review by gabor-mezei-arm — CHANGES_REQUESTED** [review]

File: library/bignum_mod_raw.c
Comment (minosgalanakis) [reply library/bignum_mod_raw.c:188]: The `m` notation is used in both mod and mod_raw, and the convention's ticket is still under design review. We should not block this pr until we decide either way. 

If a change from `N` to `m` across both modules would be required, a follow up issue will be raised and added to the technical debt.

File: library/bignum_mod_raw.h
Comment (tom-cosgrove-arm) [inline library/bignum_mod_raw.h:172]: Hmm, this change here seems to cause this PR to conflict with #6742, so I think it would be best to undo it

**Review by gabor-mezei-arm — APPROVED** [review]
The naming convention issue is handled in a separate PR.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
The `ABI-API-checking` failures can be ignored - these are due to the `mbedtls_xor()` and other functions that were added to `development` from after where this PR was based

**Comment by tom-cosgrove-arm** [comment]
Internal CI has passed except for the expect ABI-API false positive (as explained above).

**Pull request merged**
